### PR TITLE
[WIP] SCHEDULERS: Fix race condition

### PR DESCRIPTION
**Pull request by mkroening**
Fixes https://github.com/hermitcore/libhermit-rs/issues/301.

Vec::insert does not work when initializing core schedulers in the wrong order. The mutex synchronizes accesses to the map.

### CODE REVIEW

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try

**Comment by stlankes** [comment]
How do we wake-up the cores in a wrong order?

**Comment by mkroening** [comment]
> How do we wake-up the cores in a wrong order?

I am not sure, actually, but it is what happens: https://github.com/hermitcore/libhermit-rs/issues/301

We have two vCPUs. The boot vCPU might be running on a core with a high load, so maybe the second application core is faster than the first one and tries to register it's scheduler before the first one because of that. But the actual cause is just a guess.

I am pretty confident, that this fixes the issue though. Current CI runs seem to confirm that.

**Comment by mkroening** [comment]
bors try

**Comment by stlankes** [comment]
I used https://github.com/hermitcore/libhermit-rs/blob/master/src/arch/x86_64/kernel/mod.rs#L447-L449 to synchronize the wakeup process. At first a core muss call this function, before we wake up the next core.

**Comment by mkroening** [comment]
Good point. Hmm... I'll try something else then. :thinking:

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

File: src/arch/x86_64/kernel/mod.rs [hunk 1]
```diff
@@ -481,7 +481,7 @@ unsafe fn pre_init(boot_info: &'static mut BootInfo) -> ! {
 
 	BOOT_INFO = boot_info as *mut BootInfo;
 
-	if boot_info.cpu_online == 0 {
+	if intrinsics::atomic_load_acq(&(*BOOT_INFO).cpu_online) == 0 {
```
Comment (stlankes) [inline src/arch/x86_64/kernel/mod.rs:484]: Maybe you should use https://github.com/hermitcore/libhermit-rs/blob/master/src/arch/x86_64/kernel/mod.rs#L254

Comment (mkroening) [reply src/arch/x86_64/kernel/mod.rs:484]: I tried atomically loading with acquire on purpose, but this does not seem the problem. :thinking:

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
Ugh. Not being able to reproduce the problem reliably, even in CI, is a nightmare.

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try-

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
bors try

**Comment by mkroening** [comment]
Superseded by https://github.com/hermitcore/libhermit-rs/pull/509.

**Pull request closed without merging**